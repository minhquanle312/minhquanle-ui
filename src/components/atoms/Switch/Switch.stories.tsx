// Libraries
import React, { useState } from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import { CheckOutlined, CloseOutlined } from '@ant-design/icons';

// Components
import { Switch } from './Switch';
import { Button, Space } from '../index';

export default {
  title: 'Atoms/Switch',
  component: Switch,
  argTypes: {
    autoFocus: {
      name: 'autoFocus',
      defaultValue: false,
      description: 'Whether get focus when component mounted',
      table: {
        type: { summary: 'boolean' },
        defaultValue: { summary: false },
      },
      control: {
        type: 'boolean',
      },
    },
    checked: {
      name: 'checked',
      description: 'Determine whether the Switch is checked	',
      table: {
        type: { summary: 'boolean' },
        defaultValue: { summary: false },
      },
      control: {
        type: 'boolean',
      },
    },
    checkedChildren: {
      name: 'checkedChildren',
      defaultValue: false,
      description: 'The content to be shown when the state is checked	',
      table: {
        type: { summary: 'ReactNode' },
      },
      control: null,
    },
    className: {
      name: 'className',
      defaultValue: false,
      description: 'The additional class to Switch	',
      table: {
        type: { summary: 'string' },
      },
      control: null,
    },
    defaultChecked: {
      name: 'defaultChecked',
      defaultValue: false,
      description: 'Whether to set the initial state	',
      table: {
        type: { summary: 'boolean' },
        defaultValue: { summary: false },
      },
      control: {
        type: 'boolean',
      },
    },
    disabled: {
      name: 'disabled',
      defaultValue: false,
      description: 'Disable switch	',
      table: {
        type: { summary: 'boolean' },
        defaultValue: { summary: false },
      },
      control: {
        type: 'boolean',
      },
    },
    loading: {
      name: 'loading',
      defaultValue: false,
      description: 'Disable switch	',
      table: {
        type: { summary: 'boolean' },
        defaultValue: { summary: false },
      },
      control: {
        type: 'boolean',
      },
    },
    size: {
      name: 'size',
      defaultValue: 'default',
      description: 'The size of the Switch, options: `default` `small`	',
      table: {
        type: { summary: 'string' },
        defaultValue: { summary: 'default' },
      },
      control: {
        type: 'select',
        options: ['default', 'small'],
      },
    },
    unCheckedChildren: {
      name: 'unCheckedChildren',
      description: 'The content to be shown when the state is unchecked',
      table: {
        type: { summary: 'ReactNode' },
      },
      control: null,
    },
    onChange: {
      name: 'onChange',
      description: 'Trigger when the checked state is changing	',
      table: {
        type: { summary: 'function(checked: boolean, event: Event)	' },
      },
      control: null,
    },
    onClick: {
      name: 'onClick',
      description: 'Trigger when clicked	',
      table: {
        type: { summary: 'function(checked: boolean, event: Event)	' },
      },
      control: null,
    },
  },
  parameters: {
    docs: {
      description: {
        component:
          'Switching Selector.' +
          '\n### When To Use' +
          '\n' +
          '- If you need to represent the switching between two states or on-off state.' +
          '\n' +
          '- The difference between `Switch` and `Checkbox` is that `Switch` will trigger a state change directly when you toggle it, while `Checkbox` is generally used for state marking, which should work in conjunction with submit operation' +
          '\n',
      },
    },
  },
} as ComponentMeta<typeof Switch>;

// Default
const Template: ComponentStory<typeof Switch> = args => <Switch defaultChecked {...args} />;

export const Default = Template.bind({});

export const Disabled = () => {
  const [disabled, setDisabled] = useState(true);

  const toggle = () => {
    setDisabled(!disabled);
  };

  return (
    <Space direction="vertical">
      <Switch disabled={disabled} defaultChecked />
      <Button type="primary" onClick={toggle}>
        Toggle disabled
      </Button>
    </Space>
  );
};

Disabled.parameters = {
  docs: {
    description: {
      story: 'Disabled state of `Switch`.',
    },
  },
};

export const TextIcon = () => (
  <Space direction="vertical">
    <Switch checkedChildren="开启" unCheckedChildren="关闭" defaultChecked />
    <Switch checkedChildren="1" unCheckedChildren="0" />
    <Switch
      checkedChildren={<CheckOutlined />}
      unCheckedChildren={<CloseOutlined />}
      defaultChecked
    />
  </Space>
);

TextIcon.parameters = {
  docs: {
    description: {
      story: 'With text and icon.',
    },
  },
};

export const TwoSizes = () => (
  <>
    <Switch defaultChecked />
    <br />
    <Switch size="small" defaultChecked />
  </>
);

TwoSizes.parameters = {
  docs: {
    description: {
      story: '`size="small"` represents a small sized switch.',
    },
  },
};

export const Loading = () => (
  <>
    <Switch loading defaultChecked />
    <br />
    <Switch size="small" loading />
  </>
);

Loading.parameters = {
  docs: {
    description: {
      story: 'Mark a pending state of switch.',
    },
  },
};
