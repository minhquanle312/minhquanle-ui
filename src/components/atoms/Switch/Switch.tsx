import { Switch as AntdSwitch } from 'antd';

export const Switch = AntdSwitch;
