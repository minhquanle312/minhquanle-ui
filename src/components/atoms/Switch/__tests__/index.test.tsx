// Libraries
import React, { useState } from 'react';
import { render, fireEvent } from '@testing-library/react';
// Components
import { Switch } from '../Switch';

const SwitchComponent = () => {
  const [checked, setChecked] = useState(false);

  const handleChange = value => {
    setChecked(value);
  };

  return (
    <div>
      <Switch checked={checked} onChange={handleChange} />
    </div>
  );
};
describe('Switch', () => {
  it('should match snapshot', () => {
    const loadingIndicator = render(<Switch />);
    expect(loadingIndicator.container.firstChild).toMatchSnapshot();
  });
  it('render component changes state on click', () => {
    const { getByRole } = render(<SwitchComponent />);
    const switchElement = getByRole('switch');

    expect(switchElement).not.toBeChecked();

    fireEvent.click(switchElement);

    expect(switchElement).toBeChecked();
  });

  it('callback on change when click', () => {
    const handleChange = jest.fn();
    const { getByRole } = render(<Switch onChange={handleChange} />);
    const switchElement = getByRole('switch');

    fireEvent.click(switchElement);

    expect(handleChange).toHaveBeenCalledTimes(1);

    fireEvent.click(switchElement);

    expect(handleChange).toHaveBeenCalledTimes(2);
  });
});
