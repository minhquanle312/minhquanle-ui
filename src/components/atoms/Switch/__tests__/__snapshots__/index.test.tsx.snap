// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch should match snapshot 1`] = `
<button
  aria-checked="false"
  class="ant-switch css-dev-only-do-not-override-nnuwmp"
  role="switch"
  type="button"
>
  <div
    class="ant-switch-handle"
  />
  <span
    class="ant-switch-inner"
  >
    <span
      class="ant-switch-inner-checked"
    />
    <span
      class="ant-switch-inner-unchecked"
    />
  </span>
</button>
`;
